Mathlib/LinearAlgebra/Pi.lean

```diff
@@ -113,25 +115,109 @@ theorem apply_single [AddCommMonoid M] [Module R M] [DecidableEq ι] (f : (i : 
+
+theorem proj_single_same (i : ι) : (proj i).comp (single R φ i) = id :=
+  LinearMap.ext <| single_same R φ i
+
+theorem proj_single_ne (i j : ι) (h : i ≠ j) : (proj i).comp (single R φ j) = 0 :=
+  LinearMap.ext <| single_ne R φ _ _ h
+
+theorem iSup_range_single_le_iInf_ker_proj (I J : Set ι) (h : Disjoint I J) :
+    ⨆ i ∈ I, range (single R φ i) ≤ ⨅ i ∈ J, ker (proj i : (∀ i, φ i) →ₗ[R] φ i) := by
+  refine iSup_le fun i => iSup_le fun hi => range_le_iff_comap.2 ?_
+  simp only [← ker_comp, eq_top_iff, SetLike.le_def, mem_ker, comap_iInf, mem_iInf]
+  rintro b - j hj
+  rw [proj_single_ne R φ j i, zero_apply]
+  rintro rfl
+  exact h.le_bot ⟨hi, hj⟩
+
+theorem iInf_ker_proj_le_iSup_range_single {I : Finset ι} {J : Set ι} (hu : Set.univ ⊆ ↑I ∪ J) :
+    ⨅ i ∈ J, ker (proj i : (∀ i, φ i) →ₗ[R] φ i) ≤ ⨆ i ∈ I, range (single R φ i) :=
+  SetLike.le_def.2
+    (by
+      intro b hb
+      simp only [mem_iInf, mem_ker, proj_apply] at hb
+      rw [←
+        show (∑ i ∈ I, single R φ i (b i)) = b by
+          ext i
+          rw [Finset.sum_apply, ← single_same R φ i (b i)]
+          refine Finset.sum_eq_single i (fun j _ ne => single_ne _ _ _ _ ne.symm _) ?_
+          intro hiI
+          rw [single_same]
+          exact hb _ ((hu trivial).resolve_left hiI)]
+      exact sum_mem_biSup fun i _ => mem_range_self (single R φ i) (b i))
+
+theorem iSup_range_single_eq_iInf_ker_proj {I J : Set ι} (hd : Disjoint I J)
+    (hu : Set.univ ⊆ I ∪ J) (hI : Set.Finite I) :
+    ⨆ i ∈ I, range (single R φ i) = ⨅ i ∈ J, ker (proj i : (∀ i, φ i) →ₗ[R] φ i) := by
+  refine le_antisymm (iSup_range_single_le_iInf_ker_proj _ _ _ _ hd) ?_
+  have : Set.univ ⊆ ↑hI.toFinset ∪ J := by rwa [hI.coe_toFinset]
+  refine le_trans (iInf_ker_proj_le_iSup_range_single R φ this) (iSup_mono fun i => ?_)
+  rw [Set.Finite.mem_toFinset]
+
+theorem iSup_range_single [Finite ι] : ⨆ i, range (single R φ i) = ⊤ := by
+  cases nonempty_fintype ι
+  convert top_unique (iInf_emptyset.ge.trans <| iInf_ker_proj_le_iSup_range_single R φ _)
+  · rename_i i
+    exact ((@iSup_pos _ _ _ fun _ => range <| single R φ i) <| Finset.mem_univ i).symm
+  · rw [Finset.coe_univ, Set.union_empty]
+
+theorem disjoint_single_single (I J : Set ι) (h : Disjoint I J) :
+    Disjoint (⨆ i ∈ I, range (single R φ i)) (⨆ i ∈ J, range (single R φ i)) := by
+  refine
+    Disjoint.mono (iSup_range_single_le_iInf_ker_proj _ _ _ _ <| disjoint_compl_right)
+      (iSup_range_single_le_iInf_ker_proj _ _ _ _ <| disjoint_compl_right) ?_
+  simp only [disjoint_iff_inf_le, SetLike.le_def, mem_iInf, mem_inf, mem_ker, mem_bot, proj_apply,
+    funext_iff]
+  rintro b ⟨hI, hJ⟩ i
+  classical
+    by_cases hiI : i ∈ I
+    · by_cases hiJ : i ∈ J
+      · exact (h.le_bot ⟨hiI, hiJ⟩).elim
+      · exact hJ i hiJ
+    · exact hI i hiI
+
+theorem single_eq_single {a : R} :
+    (fun i : ι => (single R (fun _ : ι => R) i) a) = fun i : ι => ↑(Finsupp.single i a) :=
+  funext fun i => (Finsupp.single_eq_pi_single i a).symm
```

Comment:
This lemma strikes me as very strange; I'd argue that any user of it should be using `Finsupp.single_eq_pi_single` directly.